File: app/src/main/java/com/kickstarter/services/apiresponses/commentthreadenvelope/PageInfoEnvelope.java
package com.kickstarter.services.apiresponses.commentthreadenvelope;

import android.os.Parcelable;
import com.kickstarter.libs.qualifiers.AutoGson;
import auto.parcel.AutoParcel;

@AutoGson
@AutoParcel
public abstract class PageInfoEnvelope implements Parcelable {
    public abstract Boolean hasPreviousPage();
    public abstract Boolean hasNextPage();
    public abstract String startCursor();
    public abstract String endCursor();

    @AutoParcel.Builder
    public abstract static class Builder {
        public abstract  Builder hasPreviousPage(Boolean __);
        public abstract  Builder hasNextPage(Boolean __);
        public abstract  Builder startCursor(String __);
        public abstract  Builder endCursor(String __);
        public abstract PageInfoEnvelope build();
    }

    public static Builder builder() {
            return new AutoParcel_PageInfoEnvelope.Builder();
    }

    public abstract Builder toBuilder();
}

File: app/src/main/java/com/kickstarter/services/apiresponses/commentresponse/PageInfoEnvelope.kt
package com.kickstarter.services.apiresponses.commentresponse

import android.os.Parcelable
import kotlinx.android.parcel.Parcelize

@Parcelize
class PageInfoEnvelope private constructor(
    val hasPreviousPage: Boolean?,
    val hasNextPage: Boolean?,
    val startCursor: String?,
    val endCursor: String?
) : Parcelable {

    @Parcelize
    data class Builder(
        var hasPreviousPage: Boolean? = null,
        var hasNextPage: Boolean? = null,
        var startCursor: String? = null,
        var endCursor: String? = null
    ) : Parcelable {

        fun hasPreviousPage(hasPreviousPage: Boolean) = apply { this.hasPreviousPage = hasPreviousPage }
        fun hasNextPage(hasNextPage: Boolean) = apply { this.hasNextPage = hasNextPage }
        fun startCursor(startCursor: String) = apply { this.startCursor = startCursor }
        fun endCursor(endCursor: String) = apply { this.endCursor = endCursor }
        fun build() = PageInfoEnvelope(hasPreviousPage, hasNextPage, startCursor, endCursor)
    }

    companion object {
        fun builder() = Builder()
    }

    fun toBuilder() = Builder(this.hasPreviousPage, this.hasNextPage, this.startCursor, this.endCursor)
}
